dataservice/api/biospecimen/models.py

```diff
@@ -111,3 +131,108 @@ class Biospecimen(db.Model, Base):
     biospecimen_genomic_files = db.relationship(BiospecimenGenomicFile,
                                                 backref='biospecimen',
                                                 cascade='all, delete-orphan')
+    diagnoses = db.relationship('Diagnosis', secondary='biospecimen_diagnosis',
+                                backref=db.backref(
+                                    'biospecimens'))
+
+
+def validate_diagnosis_biospecimen(target):
+    """
+    Ensure that both the diagnosis and biospecimen
+    have the same participant
+    If this is not the case then raise DatabaseValidationError
+    """
+    from dataservice.api.errors import DatabaseValidationError
+    # Return if biospecimen_diagnosis is None
+    if not target:
+        return
+
+    # Get biospecimen and diagnosis by id
+    bsp = None
+    ds = None
+    if target.biospecimen_id and target.diagnosis_id:
+        bsp = Biospecimen.query.get(target.biospecimen_id)
+        ds = Diagnosis.query.get(target.diagnosis_id)
+
+    # If biospecimen and diagnosis doesn't exist, return and
+    # let ORM handle non-existent foreign key
+    if bsp is None or ds is None:
+        return
+
+    # Check if this diagnosis and biospecimen refer to same participant
+    if ds.participant_id != bsp.participant_id:
+        operation = 'modify'
+        target_entity = BiospecimenDiagnosis.__tablename__
+        message = (
+            ('a diagnosis cannot be linked with a biospecimen if they '
+             'refer to different participants. diagnosis {} '
+             'refers to participant {} and '
+             'biospecimen {} refers to participant {}')
+            .format(ds.kf_id,
+                    ds.participant_id,
+                    bsp.kf_id,
+                    bsp.participant_id))
+        raise DatabaseValidationError(target_entity, operation, message)
+
+
+def validate_biospecimen(target):
```

Comment:
This looks like its just validating the first diagnosis in Biospecimen.diagnoses? 

I think any time a diagnosis is appended to Biospecimen.diagnoses, you want to do the validation. See https://docs.sqlalchemy.org/en/latest/orm/events.html

Reply:
Can this validation function be removed entirely?
I believe the check for the same participant is a constraint solely on the relationship between the two entities, not the biospecimen.

Reply:
I think we could remove biospecimen_validation also diagnosis_validation but I need to test few things before we do that.